pkl/evaluator_manager_exec.go

```diff
@@ -206,5 +206,26 @@ func (e *execEvaluator) deinit() error {
 	close(e.out)
 	close(e.closed)
 	// TODO: graceful shutdown
-	return e.cmd.Process.Kill()
+	if err := e.cmd.Process.Signal(os.Interrupt); err != nil {
+		internal.Debug("Failed to interrupt process: %v", err)
+		return e.cmd.Process.Kill()
+	}
+	select {
+	case <-time.After(5 * time.Second):
+		// If the process does not exit within the timeout, kill it.
+		if killErr := e.cmd.Process.Kill(); killErr != nil {
+			internal.Debug("Failed to kill process after timeout: %v", killErr)
+			return killErr
+		}
+	case err := <-e.closed:
+		if err != nil {
+			internal.Debug("Process exited with error: %v", err)
+			// Forcefully kill the process if it exited with an error.
+			if killErr := e.cmd.Process.Kill(); killErr != nil {
+				internal.Debug("Failed to kill process after error: %v", killErr)
+				return killErr
+			}
```

Comment:
Wouldn't it already be killed at this point? The reported `err` is of a non-zero exit code, no?